Problem w/ Groovy helper function invocation

In some projects there were issues with IndexOutOfBoundsExceptions regarding calls to Groovy helper functions. For instance a call of `_.geom.findAll` with an empty list yielded this problem. It seems that this is related to reflection call mechanisms. It should be checked how we can solve this and if it is a general problem with the invocation.

Example Stacktrace:

```
eu.esdihumboldt.hale.common.align.transformation.function.TransformationException: 0
	at eu.esdihumboldt.cst.functions.groovy.GroovyRetype.execute(GroovyRetype.java:73)
	at eu.esdihumboldt.cst.ConceptualSchemaTransformer.doTypeTransformation(ConceptualSchemaTransformer.java:324)
	at eu.esdihumboldt.cst.ConceptualSchemaTransformer.transform(ConceptualSchemaTransformer.java:154)
	at eu.esdihumboldt.hale.ui.service.instance.internal.orient.OrientInstanceService$5.run(OrientInstanceService.java:464)
	at eu.esdihumboldt.hale.ui.util.io.ThreadProgressMonitor$1.run(ThreadProgressMonitor.java:121)
	at org.eclipse.jface.operation.ModalContext$ModalContextThread.run(ModalContext.java:119)
Caused by: java.lang.ArrayIndexOutOfBoundsException: 0
	at org.codehaus.groovy.reflection.ParameterTypes.coerceArgumentsToClasses(ParameterTypes.java:143)
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:322)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1207)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1016)
	at org.codehaus.groovy.runtime.callsite.PogoMetaClassSite.call(PogoMetaClassSite.java:39)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:45)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:108)
	at org.kohsuke.groovy.sandbox.impl.Checker$1.call(Checker.java:120)
	at org.kohsuke.groovy.sandbox.GroovyInterceptor.onMethodCall(GroovyInterceptor.java:21)
	at eu.esdihumboldt.util.groovy.sandbox.internal.RestrictiveGroovyInterceptor.onMethodCall(RestrictiveGroovyInterceptor.java:252)
	at org.kohsuke.groovy.sandbox.impl.Checker$1.call(Checker.java:118)
	at org.kohsuke.groovy.sandbox.impl.Checker.checkedCall(Checker.java:115)
	at org.kohsuke.groovy.sandbox.impl.Checker$checkedCall$1.callStatic(Unknown Source)
	at Script1.run(Script1.groovy:2)
	at eu.esdihumboldt.util.groovy.sandbox.DefaultGroovyService.evaluate(DefaultGroovyService.java:161)
	at eu.esdihumboldt.hale.ui.service.groovy.internal.PreferencesGroovyService.evaluate(PreferencesGroovyService.java:154)
	at eu.esdihumboldt.cst.functions.groovy.internal.GroovyUtil.evaluateAll(GroovyUtil.java:200)
	at eu.esdihumboldt.cst.functions.groovy.GroovyRetype.execute(GroovyRetype.java:60)
	... 5 more
```

--- a/cst/plugins/eu.esdihumboldt.cst.functions.groovy/plugin.xml
+++ b/cst/plugins/eu.esdihumboldt.cst.functions.groovy/plugin.xml
@@ -370,6 +370,11 @@
             forArray="false"
             implementation="eu.esdihumboldt.cst.functions.groovy.internal.AsMultiValueMetaClass">
       </metaClass>
+      <metaClass
+            for="eu.esdihumboldt.cst.functions.groovy.helper.HelperFunctions"
+            forArray="false"
+            implementation="eu.esdihumboldt.cst.functions.groovy.internal.GroovyObjectMetaClass">
+      </metaClass>
    </extension>
    <extension
          point="eu.esdihumboldt.cst.functions.groovy.helper">

--- a/cst/plugins/eu.esdihumboldt.cst.functions.groovy/src/eu/esdihumboldt/cst/functions/groovy/helper/extension/HelperFunctionsExtension.java
+++ b/cst/plugins/eu.esdihumboldt.cst.functions.groovy/src/eu/esdihumboldt/cst/functions/groovy/helper/extension/HelperFunctionsExtension.java
@@ -352,7 +352,13 @@ public class HelperFunctionsExtension implements HelperFunctionsService {
 				return null;
 			}
 			else {
-				return injectContext(catMap.get(name), context);
+				HelperFunctionOrCategory res = catMap.get(name);
+				if (res != null) {
+					return injectContext(res, context);
+				}
+				else {
+					return null;
+				}
 			}
 		}
 	}

--- a/cst/plugins/eu.esdihumboldt.cst.functions.groovy/src/eu/esdihumboldt/cst/functions/groovy/internal/GroovyObjectMetaClass.java
+++ b/cst/plugins/eu.esdihumboldt.cst.functions.groovy/src/eu/esdihumboldt/cst/functions/groovy/internal/GroovyObjectMetaClass.java
@@ -1,0 +1,74 @@
+/*
+ * Copyright (c) 2017 wetransform GmbH
+ * 
+ * All rights reserved. This program and the accompanying materials are made
+ * available under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation, either version 3 of the License,
+ * or (at your option) any later version.
+ * 
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this distribution. If not, see <http://www.gnu.org/licenses/>.
+ * 
+ * Contributors:
+ *     wetransform GmbH <http://www.wetransform.to>
+ */
+
+package eu.esdihumboldt.cst.functions.groovy.internal;
+
+import groovy.lang.DelegatingMetaClass;
+import groovy.lang.GroovyObject;
+import groovy.lang.MetaClass;
+import groovy.lang.MissingMethodException;
+
+/**
+ * Metaclass that prefers delegating method calls to {@link GroovyObject}s
+ * instead of trying to invoke the method in the meta class first.
+ * 
+ * This allows the {@link GroovyObject} overriding implementations in the meta
+ * class, for instance the default methods that Groovy adds to all objects like
+ * <code>findAll</code> or <code>each</code>.
+ * 
+ * Attention: If the {@link GroovyObject} itself delegates calls to the meta
+ * class there is a loop resulting in a stack overflow. Thus select careful for
+ * which classes/objects you use this metaclass.
+ * 
+ * @author Simon Templer
+ */
+public class GroovyObjectMetaClass extends DelegatingMetaClass {
+
+	/**
+	 * @see DelegatingMetaClass#DelegatingMetaClass(Class)
+	 */
+	public GroovyObjectMetaClass(@SuppressWarnings("rawtypes") Class theClass) {
+		super(theClass);
+		initialize();
+	}
+
+	/**
+	 * @see DelegatingMetaClass#DelegatingMetaClass(MetaClass)
+	 */
+	public GroovyObjectMetaClass(MetaClass delegate) {
+		super(delegate);
+		initialize();
+	}
+
+	@Override
+	public Object invokeMethod(Object object, String methodName, Object[] arguments) {
+		if (object instanceof GroovyObject) {
+			// Prefer the GroovyObject implementation over an implementation in
+			// the meta class.
+			// Allows delegating Groovy default methods like findAll to the
+			// Groovy object.
+			try {
+				return ((GroovyObject) object).invokeMethod(methodName, arguments);
+			} catch (MissingMethodException | StackOverflowError e) {
+				// recover calling original meta class
+				return super.invokeMethod(object, methodName, arguments);
+			}
+		}
+		else {
+			return super.invokeMethod(object, methodName, arguments);
+		}
+	}
+
+}

